cmake/AddCustomXXXTimestamp.cmake

```diff
@@ -0,0 +1,145 @@
+macro(_parse_arguments_with_multi_hack ORIGINAL_COMMAND_LINE)
+    # cmake_parse_arguments() put all the MULTIS in a single variable; you
+    # lose the ability to see for example multiple COMMANDs. To be able to
+    # passthrough multiple MULTIS, we add a marker after every MULTI. This
+    # allows us to reassemble the correct amount again before giving it to
+    # the wrapped command with _reassemble_command_line().
+
+    set(COMMAND_LINE "${ORIGINAL_COMMAND_LINE}")
+
+    foreach(MULTI IN LISTS MULTIS)
+        string(REPLACE "${MULTI}" "${MULTI};:::" COMMAND_LINE "${COMMAND_LINE}")
+    endforeach(MULTI)
+
+    cmake_parse_arguments(PARAM "${OPTIONS}" "${SINGLES}" "${MULTIS}" ${COMMAND_LINE})
+endmacro()
+
+macro(_reassemble_command_line)
+    # Reassemble the command line as we original got it.
+    set(NEW_COMMAND_LINE ${PARAM_UNPARSED_ARGUMENTS})
+
+    foreach(OPTION IN LISTS OPTIONS)
+        if (PARAM_${OPTION})
+            list(APPEND NEW_COMMAND_LINE "${OPTION}")
+        endif (PARAM_${OPTION})
+    endforeach(OPTION)
+
+    foreach(SINGLE IN LISTS SINGLES)
+        if (PARAM_${SINGLE})
+            list(APPEND NEW_COMMAND_LINE "${SINGLE}" "${PARAM_${SINGLE}}")
+        endif (PARAM_${SINGLE})
+    endforeach(SINGLE)
+
+    foreach(MULTI IN LISTS MULTIS)
+        if (PARAM_${MULTI})
+            # Replace our special marker with the name of the MULTI again. This
+            # restores for example multiple COMMANDs again.
+            string(REPLACE ":::" "${MULTI}" PARAM_${MULTI} "${PARAM_${MULTI}}")
+            list(APPEND NEW_COMMAND_LINE "${PARAM_${MULTI}}")
+        endif (PARAM_${MULTI})
+    endforeach(MULTI)
+endmacro()
+
+# Generated files can be older than their dependencies, causing useless
```

Comment:
Why is this necessary? Under what circumstances can generated files be older than their dependencies? That doesn't make sense at all.

Reply:
When a dependency is modified (like the cmake script used to generate a file), the change may have no effect to the generated file, because `configure_file()` basically creates a temp file and updates the generated file only if the temp file is different.
Same happens with `strgen` and `settingsgen`, they update the table headers only if they have changed.
So yes, generated files can be older than their dependencies :)